components/director/internal/domain/formation/service.go

```diff
@@ -180,8 +195,239 @@ func (s *service) UnassignFormation(ctx context.Context, tnt, objectID string, o
+	}
+
+	targetTenant, err := s.ensureSameTargetTenant(in)
+	if err != nil {
+		return errors.Wrap(err, "while ensuring input is valid")
+	}
+
+	if err = s.repo.DeleteForTargetTenant(ctx, tenantID, targetTenant); err != nil {
+		return errors.Wrap(err, "while deleting the Assignments")
+	}
+
+	if err = s.RemoveAssignedScenarios(ctx, in); err != nil {
+		return errors.Wrap(err, "while unassigning scenario from runtimes")
+	}
+
+	return nil
+}
+
+// MergeScenariosFromInputLabelsAndAssignments merges all the scenarios that are part of the resource labels (already added + to be added with the current operation)
+// with all the scenarios that should be assigned based on ASAs.
+func (s *service) MergeScenariosFromInputLabelsAndAssignments(ctx context.Context, inputLabels map[string]interface{}, runtimeID string) ([]interface{}, error) {
+	scenariosSet := make(map[string]struct{})
+
+	scenariosFromAssignments, err := s.GetScenariosFromMatchingASAs(ctx, runtimeID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while getting scenarios for selector labels")
+	}
+
+	for _, scenario := range scenariosFromAssignments {
+		scenariosSet[scenario] = struct{}{}
+	}
+
+	scenariosFromInput, isScenarioLabelInInput := inputLabels[model.ScenariosKey]
+
+	if isScenarioLabelInInput {
+		scenariosFromInputInterfaceSlice, ok := scenariosFromInput.([]interface{})
+		if !ok {
+			return nil, apperrors.NewInternalError("while converting scenarios label to an interface slice")
+		}
+
+		for _, scenario := range scenariosFromInputInterfaceSlice {
+			scenariosSet[fmt.Sprint(scenario)] = struct{}{}
+		}
+	}
+
+	scenarios := make([]interface{}, 0)
+	for k := range scenariosSet {
+		scenarios = append(scenarios, k)
+	}
+	return scenarios, nil
+}
+
+// GetScenariosFromMatchingASAs gets all the scenarios that should be added to the runtime based on the matching Automatic Scenario Assignments
+// In order to do that, the ASAs should be searched in the caller tenant as this is the tenant that modifies the runtime and this is the tenant that the ASA
+// produced labels should be added to.
+func (s *service) GetScenariosFromMatchingASAs(ctx context.Context, runtimeID string) ([]string, error) {
+	tenantID, err := tenant.LoadFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	scenariosSet := make(map[string]struct{})
+
+	scenarioAssignments, err := s.repo.ListAll(ctx, tenantID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while listinng Automatic Scenario Assignments in tenant: %s", tenantID)
+	}
+
+	matchingASAs := make([]*model.AutomaticScenarioAssignment, 0, len(scenarioAssignments))
+	for _, scenarioAssignment := range scenarioAssignments {
+		matches, err := s.isASAMatchingRuntime(ctx, scenarioAssignment, runtimeID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "while checkig if asa matches runtime with ID %s", runtimeID)
+		}
+		if matches {
+			matchingASAs = append(matchingASAs, scenarioAssignment)
+		}
+	}
+
+	for _, sa := range matchingASAs {
```

Comment:
I know this is not your code, but is this loop actually needed? We can change the loop below to iterate through the `matchingASAs` and append the `sa.ScenarioName`. Also, the `scenariosSet` variable is not needed at all.

Reply:
Looks to me as if it's been copied from the `MergeScenariosFromInputLabelsAndAssignments` method above and just adapted the variable names and in the current method the `scenarioSet` and the loop are not needed.